implement a 'look' command like cpanm

tartansandal opened the pull request.

thaljef commented on lib/Pinto/Shell.pm:
```diff
@@ -0,0 +1,63 @@
+
+    my $shell = $self->shell;
+
+    my $cwd_guard = cwd_guard( $self->work_dir );
```
That worked out quite nicely, don't you think?

tartansandal replied:
Lovely :-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

On 6 February 2015 at 16:32, Jeffrey Ryan Thalhammer <
notifications@github.com> wrote:

> In lib/Pinto/Shell.pm
> <https://github.com/thaljef/Pinto/pull/184#discussion_r24223481>:
>
> > +has work_dir => (
> > +    is       => 'ro',
> > +    isa      => Dir,
> > +    default  => sub { $_[0]->unpacker->unpack },
> > +    init_arg => undef,
> > +    lazy     => 1,
> > +);
> > +
> > +#------------------------------------------------------------------------------
> > +
> > +sub spawn {
> > +    my ($self) = @_;
> > +
> > +    my $shell = $self->shell;
> > +
> > +    my $cwd_guard = cwd_guard( $self->work_dir );
>
> That worked out quite nicely, don't you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184/files#r24223481>.
>

thaljef commented on lib/App/Pinto/Command/look.pm:
```diff
@@ -0,0 +1,85 @@
+
+=head1 COMMAND ARGUMENTS
+
+Arguments are the targets you wish to unpin.  Targets can be
```
"unpin" -> "look at"

tartansandal replied:
he he. nice catch.


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

On 2 March 2015 at 08:12, Jeffrey Ryan Thalhammer <notifications@github.com>
wrote:

> In lib/App/Pinto/Command/look.pm
> <https://github.com/thaljef/Pinto/pull/184#discussion_r25571513>:
>
> > +
> > +__END__
> > +
> > +
> > +=head1 SYNOPSIS
> > +
> > +  pinto --root=REPOSITORY_ROOT look [OPTIONS] TARGET ...
> > +
> > +=head1 DESCRIPTION
> > +
> > +Unpack one or more distributions and open directories with your shell.
> > +Handy if you want to manually review a distribution before use.
> > +
> > +=head1 COMMAND ARGUMENTS
> > +
> > +Arguments are the targets you wish to unpin.  Targets can be
>
> "unpin" -> "look at"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184/files#r25571513>.
>

thaljef commented on lib/Pinto/Action/Look.pm:
```diff
@@ -0,0 +1,75 @@
+# ABSTRACT: Force a package to stay in a stack
```
copy-n-pasted

thaljef commented on lib/Pinto/Action/Look.pm:
```diff
@@ -0,0 +1,75 @@
+
+has shell => (
+    is       => 'ro',
+    isa      => Str,
```
Should probably use the `File` type from `Pinto::Types` here. Be sure to set `coerce => 1`

thaljef commented on lib/Pinto/Shell.pm:
```diff
@@ -0,0 +1,63 @@
+use MooseX::Types::Moose qw(Str);
+
+use Pinto::Types qw(File Dir);
+use Pinto::Util qw(debug throw);
```
I don't see `debug` or `throw` being used anywhere.

thaljef commented:
This looks really good, but we can't release it until it works reasonably with a remote repository. There are two choices for that:

1. Implement support for remote repositories by creating a subclass of `Pinto::Remote::Action` that fetches the remote tarball first (see `Pinto::Remote::Action::Install`l for example).

1. Make `App::Pinto::Command::look::execute()` bail out with an appropriate error message if the repository is remote (i.e. starts with /https?/). Then we'll add support for remote repositories later.

Either is fine with me.

tartansandal commented:
Okay. Implemented 3 patches to fix copy-n-paste errors, 1 to make SHELL a
File <doh>, and 1 to die if the remote repository is remote (inspired by
the 'init' command). Opening distros from remote repositories makes good
sense, but I'm a little short on time ATM.

I'm using next::method() from mro 'cause its been so long and I can't
remember how to use SUPER anymore. Really wanted to use a 'before'
modifier, but App::Cmd does not seem to role that way. Added MRO::compat so
users of RHEL5 will be happier. Opinions most welcome ;-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

On 2 March 2015 at 09:13, Jeffrey Ryan Thalhammer <notifications@github.com>
wrote:

> This looks really good, but we can't release it until it works reasonably
> with a remote repository. There are two choices for that:
>
>    1.
>
>    Implement support for remote repositories by creating a subclass of
>    Pinto::Remote::Action that fetches the remote tarball first (see
>    Pinto::Remote::Action::Installl for example).
>    2.
>
>    Make App::Pinto::Command::look::execute() bail out with an appropriate
>    error message if the repository is remote (i.e. starts with /https?/). Then
>    we'll add support for remote repositories later.
>
> Either is fine with me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184#issuecomment-76634331>.
>

thaljef commented:
App::Cmd isn't based on Moo/Moose, so method modifiers are not an option there. I don't have anything against MRO (DBIx::Class is already using it). But for the sake of consistency, we have to choose one way or the other.  Currently we have no `next::method()` calls in the code but two of these:

```
App/Pinto/Command/help.pm:24:    my $rv = $self->SUPER::execute( $opts, $args );
App/Pinto/Command/nop.pm:29:    $self->SUPER::validate_args( $opts, $args );
```

tartansandal commented:
I reckon we stick with the status quo -- now that I've been reminded of how
to use SUPER once again :-)

We don't have complex/multiple inheritance under App::Cmd so the whole MRO
thing is not important.

Will get onto a patch for this an hour or so.


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

On 3 March 2015 at 17:03, Jeffrey Ryan Thalhammer <notifications@github.com>
wrote:

> App::Cmd isn't based on Moo/Moose, so method modifiers are not an option
> there. I don't have anything against MRO (DBIx::Class is already using it).
> But for the sake of consistency, we have to choose one way or the other.
> Currently we have no next::method() calls in the code but two of these:
>
> App/Pinto/Command/help.pm:24:    my $rv = $self->SUPER::execute( $opts, $args );
> App/Pinto/Command/nop.pm:29:    $self->SUPER::validate_args( $opts, $args );
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184#issuecomment-76892151>.
>

thaljef commented:
> Will get onto a patch for this an hour or so.

Already done: bb113a

thaljef commented:
I already merged this, but I just realized this command needs at `stack` option too.  Whenever a command accepts package names like Foo::Bar as a target, then it needs to know which stack to resolve that against, because every stack could have a different version of Foo::Bar.  This is an important feature of Pinto's behavior. I'll try to fix this so you can see what I mean.

tartansandal commented:
Ah yes, that makes sense. Probably true of the sign command too.
On 03/03/2015 5:32 pm, "Jeffrey Ryan Thalhammer" <notifications@github.com>
wrote:

> I already merged this, but I just realized this command needs at stack
> option too. Whenever a command accepts package names like Foo::Bar as a
> target, then it needs to know which stack to resolve that against, because
> every stack could have a different version of Foo::Bar. This is an
> important feature of Pinto's behavior. I'll try to fix this so you can see
> what I mean.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184#issuecomment-76894340>.
>

tartansandal commented:
Actually, ignore my last comment. When we sign, we sign all the authors
distributions. Hence the clone command to get a minimal set to sign.
On 03/03/2015 5:41 pm, "Kahlil Hodgson" <kahlil.hodgson@dealmax.com.au>
wrote:

> Ah yes, that makes sense. Probably true of the sign command too.
> On 03/03/2015 5:32 pm, "Jeffrey Ryan Thalhammer" <notifications@github.com>
> wrote:
>
>> I already merged this, but I just realized this command needs at stack
>> option too. Whenever a command accepts package names like Foo::Bar as a
>> target, then it needs to know which stack to resolve that against, because
>> every stack could have a different version of Foo::Bar. This is an
>> important feature of Pinto's behavior. I'll try to fix this so you can see
>> what I mean.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/thaljef/Pinto/pull/184#issuecomment-76894340>.
>>
>

thaljef commented:
> When we sign, we sign all the authors distributions.

Different versions of a package could be owned by different authors. So I think the `sign` command still needs to resolve a package target to a specific distribution.

thaljef commented:
Never mind that last comment. I see now that it signs *every* distribution by *every* author. So I'm going to delete the last part of this thread.

thaljef commented:
I rewrote some of this in 09c626bd. Mostly I just re-arranged things. But I did make one interface change by removing the `--shell` command line option in favor of a `PINTO_SHELL` environment variable, which is consistent with other parts of the interface, such as `PINTO_EDITOR` and `PINTO_PAGER`. Thanks to you, this is a very nice addition to Pinto :smiley_cat:

tartansandal commented:
Nice reorg. "Look"ing good ;-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

On 4 March 2015 at 05:14, Jeffrey Ryan Thalhammer <notifications@github.com>
wrote:

> I rewrote some of this in 09c626b
> <https://github.com/thaljef/Pinto/commit/09c626bd6c2b3bbec27b8f103d214a486a201ffe>.
> Mostly I just re-arranged things. But I did make one interface change by
> removing the --shell command line option in favor of a PINTO_SHELL
> environment variable, which is consistent with other parts of the
> interface, such as PINTO_EDITOR and PINTO_PAGER. Thanks to you, this is a
> very nice addition to Pinto [image: :smiley_cat:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thaljef/Pinto/pull/184#issuecomment-77001599>.
>
